src/main/java/dev/mccue/resolve/maven/MavenRepository.java

```diff
@@ -11,5 +31,172 @@ public static boolean isSnapshot(String version) {
                 || SNAPSHOT_TIMESTAMP.matcher(version).matches();
     }
 
+    public static String toBaseVersion(String version) {
+        switch (version) {
+            //TODO
+
+        }
+        return version;
+    }
+
+    public static Optional<String> mavenVersioning (
+        SnapshotVersioning snapshotVersioning,
+        Classifier classifier,
+        Extension extension
+    ) {
+        return snapshotVersioning
+            .snapshotVersions()
+            .stream()
+            .filter((SnapshotVersion v) -> {
+                return (v.classifier().equals(classifier) || v.classifier().equals(new Classifier("*"))) && 
+                       (v.extension().equals(extension) || v.extension().equals(new Extension("*")));
+            })
+            .findFirst()
+            .map(v -> v.value());//I think this is right, coursier uses an extra .filter?
+    }
+
+    private static String dirModuleName(Module module, Boolean sbtAttrStub) {
+        if (sbtAttrStub) {
+            var name = module.name().value();
+            return name;
+        } else {
+            return module.name().value();
+        }
+    }
+
+    private static Project parseRawPomSax(String str) throws SaxParsingException {
+        try {
+            return Utilities.xmlParseSax(str, new PomParser()).project();
+        } catch (SaxParsingException e) { throw e;
+        }
+    }
+
+    private static String actualRoot(String root) {
+        if (root.endsWith("/")) 
+            return root.substring(0, root.length() - 1);
+        else 
+            return root;
+    }    
+
+    public static MavenRepository apply(String root) {
+        return new MavenRepository(actualRoot(root));
+    }
+    public static MavenRepository apply(String root, Optional<Authentication> authentication) {
+        return new MavenRepository(root, authentication);
+    }
```

Comment:
Apply methods are interesting. They mostly exist in scala as syntax sugar. It would make sense potentially to just make these the behavior of the constructor.